### PR TITLE
ForceRefresh()/ForceRefreshAsync() should return with a result

### PR DESCRIPTION
**Pull request by adams85**
### Describe the purpose of your pull request

Breaking changes:
* Changes `IConfigCatClient.RefreshConfig`/`RefreshConfigAsync` to return information about the success of the operation.

- [x] I have covered the applied changes with automated tests.
- [x] I have executed the full automated test set against my changes.
- [x] I have validated my changes against all supported platform versions.
- [x] I have read and accepted the [contribution agreement](https://github.com/configcat/legal/blob/main/contribution-agreement.md).

### CODE REVIEW

File: src/ConfigCatClient/ConfigService/ConfigServiceBase.cs [hunk 1]
```diff
@@ -112,16 +112,18 @@ protected ProjectConfig RefreshConfigCore(ProjectConfig latestConfig)
...
         }
 
-        public virtual async Task RefreshConfigAsync()
+        public virtual async Task<RefreshResult> RefreshConfigAsync()
         {
             if (!IsOffline)
             {
                 var latestConfig = await this.ConfigCache.GetAsync(this.CacheKey).ConfigureAwait(false);
                 await RefreshConfigCoreAsync(latestConfig).ConfigureAwait(false);
+                return RefreshResult.Success();
```
Comment (z4kn4fein) [inline src/ConfigCatClient/ConfigService/ConfigServiceBase.cs:121]: I think we should bubble up errors from the `ConfigFetcher` too and set the error of the result accordingly. WDYT?

In other SDKs, I wrapped the returning `ProjectConfig` into a `FetchResponse` object which is supposed to track the status, the actual config in case of 200 OK, and the error if there was any. 
(I know this will kill a lot of tests 😿)

File: src/ConfigCatClient/FetchResult.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            return new FetchResult(
+                config ?? throw new ArgumentNullException(nameof(config)),
+                errorMessage ?? throw new ArgumentNullException(nameof(errorMessage)),
+                errorException);
+        }
+
+        private readonly object errorMessageOrToken;
+
+        public FetchResult(ProjectConfig config, object errorMessageOrToken, Exception errorException)
```
Comment (z4kn4fein) [inline src/ConfigCatClient/FetchResult.cs:31]: This is a bit confusing for me, wouldn't be simpler expecting a `string?` here as `errorMessage`? So the `ErrorMessage` prop could be set right there, just as the prop `IsFailure => ErrorMessage is not null;`
Also, `SuccessToken` and `NotModifiedToken` could be deleted this way.
Let me know if I misunderstood something!

Comment (adams85) [reply src/ConfigCatClient/FetchResult.cs]: There's a story behind this: at first there were an `IsSuccess` and an `IsNotModified` property as well because I wanted to preserve the status information. Later on I realized that those properties are kinda useless because an `IsFailure` property is enough to implement the logic. So I deleted the former ones and ended up with this mess, where you're remark makes perfect sense... :)

However, it still seems better to preserve the status information so we can tell whether the fetch result was success or not modified. Although the library code doesn't need this information *currently* but it can be useful for tests and debugging.

So, how about [this](https://github.com/configcat/.net-sdk/pull/54/commits/193e356697fe2a47507ee5142887e3d839d939aa#diff-d6587e58e188cc3aab8f43dbf2803a2e9aa3191b82f4f65173b5001a918a9fcbR36-R38)?

Comment (endret) [reply src/ConfigCatClient/FetchResult.cs]: And what do you think about introducing a `FetchResultType` enum? With the static factory methods, you can set only an implicit way.

Comment (adams85) [reply src/ConfigCatClient/FetchResult.cs]: You mean replacing the `IsXXX` properties with something like
```csharp
public FetchResultType Type => this.errorMessageOrToken switch
{
    null => FetchResultType.Success,
    string => FetchResultType.Failure,
    _ => FetchResultType.NotModified
};
```
or you'd even do `public FetchResultType Type { get; }`, in other words, would store the type information directly?

I agree that introducing an enum could improve the API somewhat, however this is an internal type so it's probably not that crucial. (I think that the proposed API is not that terrible because I saw [similar designs in the framework](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.authenticateresult?view=aspnetcore-6.0#properties) itself.)

However, to keep the struct as slim as possible, I'd avoid adding another field if it's not necessarily needed. And as I demonstrated, it's not necessary since we can encode the status information with that little trick involving `errorMessageOrToken`.

> With the static factory methods, you can set only an implicit way.

I'm afraid I'm not completely understand this.

`FetchResult` should actually be a discriminated union. Unfortunately, C# doesn't have this concept (yet), so we can just emulate this. Hence the static factory methods. I know of several ways of emulating DUs but in all cases you need static factories if you want to provide a clean API. You do maybe?

Comment (endret) [reply src/ConfigCatClient/FetchResult.cs]: You can store directly, and the class's public factory methods can set this enum only with the private constructor, nothing more. The enum could be nested, and you can use expose your properties similar way with one line like:
`public bool IsSuccess => this.type == FetchResultType.Success`

Comment (adams85) [reply src/ConfigCatClient/FetchResult.cs]: We could, of course, do this but I fail to see what we gain from it. As mentioned above, the proposed solution is already capable of storing the type information without introducing an enum + adding an extra field.

If we introduce an enum, then we should use that in the public API of the type (I mean `public FetchResultType Type { get; }`) so at least the API is improved a bit. Otherwise it's just an implementation detail which is completely irrelevant to the consumers of the type, i.e. a matter of preference in my eyes. (At least, I think my version is still pretty easy to follow, so it may not fall into the category of unacceptable hacks.)

To sum it up, we have two options to choose from:
1. Additional enum and field for extra clarity + a single `Type` property instead of multiple `IsXXX` ones.
2. The current proposal focusing on an economic design.

How about leaving the decision to @z4kn4fein, the initiator of the discussion?

Comment (z4kn4fein) [reply src/ConfigCatClient/FetchResult.cs]: I'm ok with the current solution

**Review by z4kn4fein — APPROVED** [review]

File: src/ConfigCatClient/FetchResult.cs (hunk 2 above)
Comment (adams85) [reply src/ConfigCatClient/FetchResult.cs]: @endret Are you ok with it as well? Can we go ahead and merge this PR?

Comment (endret) [reply src/ConfigCatClient/FetchResult.cs]: ok.

(to tell you the truth the enum-based solution is clearer for me than this)

**Review by endret — APPROVED** [review]
